File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/python/fenwick-tree.py
"""
Python Implementation: Fenwick Tree (Binary Indexed Tree)
Invariant: Prefix Sum O(log N), Update O(log N), Space O(N)
"""
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class FenwickTreeBinaryIndexedTree(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    # TODO: Implement core operational method
    def insert(self, value: T) -> None:
        self._items.append(value)

    def __len__(self) -> int:
        return len(self._items)

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/javascript/fenwick-tree.js
/**
 * JavaScript Implementation: Fenwick Tree (Binary Indexed Tree)
 * Invariant: Prefix Sum O(log N), Update O(log N), Space O(N)
 */
export class FenwickTreeBinaryIndexedTree {
  constructor() {
    this.items = [];
  }

  // TODO: Implement verified invariant
  insert(val) {
    this.items.push(val);
  }

  get size() {
    return this.items.length;
  }
}
